datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -467,6 +468,200 @@ impl LogicalPlan {
         self.with_new_exprs(self.expressions(), inputs.to_vec())
     }
 
+    /// Recomputes schema and type information for this LogicalPlan if needed.
+    ///
+    /// Some `LogicalPlan`s may need to recompute their schema if the number or
+    /// type of expressions have been changed (for example due to type
+    /// coercion). For example [`LogicalPlan::Projection`]s schema depends on
+    /// its expressions.
+    ///
+    /// Some `LogicalPlan`s schema is unaffected by any changes to their
+    /// expressions. For example [`LogicalPlan::Filter`] schema is always the
+    /// same as its input schema.
+    ///
+    /// # Return value
+    /// Returns an error if there is some issue recomputing the schema.
+    ///
+    /// # Notes
+    ///
+    /// * Does not recursively recompute schema for input (child) plans.
+    pub fn recompute_schema(self) -> Result<Self> {
```

Comment:
All this logic is the same as `new_with_expr`, even some questionable code like for Filter and Union

Reply:
I have a question which may be stupid, what's the difference between this and getting the inputs and exprs of the LogicalPlan and pass them as params to `new_with_expr`? 🤔

Reply:
The key difference is that to use `new_with_expr` it requires providing a copy (`Vec<Expr>` and `Vec<LogicalPlan>`) where as this function can be used after modifying the Exprs (or children) via methods such as `map_children` and `map_expressions`

Avoiding those copies is a key part of improve planning performance in PRs like https://github.com/apache/datafusion/pull/10356 (the change is basically to stop copying)

Reply:
I pushed bd7b62f2aa76e33a4fcba7cffaf02920114192bb to try and clarify